### PR TITLE
delete CheCluster when CheInstallation is deleted

### PR DESCRIPTION
**Pull request by xcoulon**
relying on the ownerRef does not work because the CheCluster
resource also has a finalizer which is handled by the che
operator, but when the CheInstallation resource is deleted,
the 'toolchain-workspaces' namespace is also deleted (by
ownership), which in turn removes the Che operator and if
this happens before the CheCluster resource was handled,
then the latter still exists and the parent namespace
is stuck in a 'Terminating' state.

Also, when the namespace is in 'Terminating' state (if
an admin deleted it manually), we delete the CheCluster
for the same reason.

Bonus: a test which was previously commented out is now
part of the E2e testsuite ;)

Fixes https://issues.redhat.com/browse/CRT-464

Signed-off-by: Xavier Coulon <xcoulon@redhat.com>

### CODE REVIEW

File: pkg/controller/cheinstallation/cheinstallation_controller.go [hunk 1]
```diff
@@ -109,14 +117,37 @@ func (r *ReconcileCheInstallation) Reconcile(request reconcile.Request) (reconci
...
+			return reconcile.Result{}, r.setStatusCheInstallationTerminating(cheInstallation, "deleting CheCluster resource")
+		} else {
+			// CheCluster resource as already deleted, we can now remove the finalizer on the CheInstallation
+			util.RemoveFinalizer(cheInstallation, toolchainv1alpha1.FinalizerName)
+			if err := r.client.Update(context.Background(), cheInstallation); err != nil {
+				return reconcile.Result{}, r.wrapErrorWithStatusUpdate(reqLogger, cheInstallation, r.setStatusCheInstallationTerminating, err, "failed to remove finalizer")
+			}
+			return reconcile.Result{}, nil
+		}
+	}
```
Comment (MatousJobanek) [inline pkg/controller/cheinstallation/cheinstallation_controller.go]: shouldn't we return in case the `CheInstallation` is being deleted but the finalizer is not there? Otherwise, the code will keep going and start checking the resources for their existance again...

File: pkg/controller/cheinstallation/cheinstallation_controller.go [hunk 2]
```diff
@@ -109,14 +117,37 @@ func (r *ReconcileCheInstallation) Reconcile(request reconcile.Request) (reconci
 	reqLogger.Info("Reconciling CheInstallation")
 
 	cheInstallation := &v1alpha1.CheInstallation{}
-	err := r.client.Get(context.TODO(), request.NamespacedName, cheInstallation)
+	err := r.client.Get(context.TODO(), types.NamespacedName{
+		Name: InstallationName,
+	}, cheInstallation)
```
Comment (MatousJobanek) [inline pkg/controller/cheinstallation/cheinstallation_controller.go]: May I know why did you change it to the fixed name instead of using the one from the request?

File: pkg/controller/cheinstallation/cheinstallation_controller.go [hunk 3]
```diff
@@ -167,8 +212,10 @@ func (r *ReconcileCheInstallation) ensureCheNamespace(logger logr.Logger, cheIns
 				return false, err
 			}
 			if ns.Status.Phase != v1.NamespaceActive {
-				logger.Info("Namespace is not in active state", "namespace", ns.Name, "phase", ns.Status.Phase)
-				return true, nil // requeue until the namespace is active
+				logger.Info("Namespace is not in active state - deleting remaining CheCluster resource", "namespace", ns.Name, "phase", ns.Status.Phase)
+				// return 'true' in any case, as we don't want to continue with the current reconciliation loop
+				_, err = r.ensureCheClusterDeletion(logger, cheInstallation)
```
Comment (MatousJobanek) [inline pkg/controller/cheinstallation/cheinstallation_controller.go]: do we need this?

File: pkg/controller/cheinstallation/cheinstallation_controller.go [hunk 4]
```diff
@@ -227,57 +275,76 @@ func (r *ReconcileCheInstallation) ensureWatchCheCluster() (bool, error) {
...
+	}, cluster); err != nil {
+		if errors.IsNotFound(err) {
+			logger.Info("CheCluster already deleted", "CheCluster.Namespace", cheInstallation.Spec.CheOperatorSpec.Namespace, "CheCluster.Name", CheClusterName)
+			return false, nil
 		}
+		logger.Info("Unexpected error while creating a CheCluster for Che", "CheCluster.Namespace", cluster.Namespace, "CheCluster.Name", cluster.Name)
+		return false, err
+		// } else if cluster.DeletionTimestamp != nil {
+		// 	logger.Info("CheCluster already deleted", "CheCluster.Namespace", cheInstallation.Spec.CheOperatorSpec.Namespace, "CheCluster.Name", CheClusterName)
+		// 	return false, nil
```
Comment (MatousJobanek) [inline pkg/controller/cheinstallation/cheinstallation_controller.go]: zombie? :ghost:

File: pkg/controller/cheinstallation/cheinstallation_controller.go [hunk 5]
```diff
@@ -227,57 +275,76 @@ func (r *ReconcileCheInstallation) ensureWatchCheCluster() (bool, error) {
...
+	}
+	logger.Info("Deleted CheCluster for Che", "CheCluster.Namespace", cluster.Namespace, "CheCluster.Name", cluster.Name)
```
Comment (MatousJobanek) [inline pkg/controller/cheinstallation/cheinstallation_controller.go]:
```suggestion
	logger.Info("Deleting CheCluster for Che", "CheCluster.Namespace", cluster.Namespace, "CheCluster.Name", cluster.Name)
```

File: pkg/controller/cheinstallation/cheinstallation_controller_test.go [hunk 6]
```diff
@@ -475,6 +481,49 @@ func TestReconcile(t *testing.T) {
...
+				fmt.Printf("deleting resource of type '%T'\n", obj)
+				if obj, ok := obj.(*orgv1.CheCluster); ok {
+					// mark namespaces as deleted...
+					deletionTS := metav1.NewTime(time.Now())
+					obj.SetDeletionTimestamp(&deletionTS)
+					// ... but replace them in the fake client cache yet instead of deleting them
+					return cl.Client.Update(ctx, obj)
+				}
+				return cl.Client.Delete(ctx, obj, opts...)
+			}
```
Comment (MatousJobanek) [inline pkg/controller/cheinstallation/cheinstallation_controller_test.go]: do we need to really care about the timestamp - couldn't we just let the client delete the resource? To verify that in the `// then` part, we would just check if the resource is gone. And we should also check that the finalizer is gone - in the similar way as it is done for example done for MUR GC: https://github.com/codeready-toolchain/host-operator/blob/master/pkg/controller/masteruserrecord/masteruserrecord_controller_test.go#L497-L519

File: pkg/controller/cheinstallation/cheinstallation_controller_test.go [hunk 7]
```diff
@@ -306,7 +307,8 @@ func TestReconcile(t *testing.T) {
 			AssertThatOperatorGroup(t, cheOperatorNS, OperatorGroupName, cl).Exists()
 			AssertThatSubscription(t, cheOperatorNS, SubscriptionName, cl).Exists()
 			AssertThatCheCluster(t, cheCluster.Namespace, cheCluster.Name, cl).Exists()
-			AssertThatCheInstallation(t, cheInstallation.Namespace, cheInstallation.Name, cl).HasNoCondition()
+			AssertThatCheInstallation(t, cheInstallation.Namespace, cheInstallation.Name, cl).
+				HasConditions(Installing("Status is unknown for CheCluster 'codeready-workspaces'"))
```
Comment (MatousJobanek) [inline pkg/controller/cheinstallation/cheinstallation_controller_test.go]: could you please also add a check (maybe also for the other test-cases when the CheCluster has been created) that the the finalizer (in the `CheInstallation` resource) has been added?

File: test/e2e/toolchain_test.go [hunk 8]
```diff
@@ -54,29 +57,33 @@ func TestToolchain(t *testing.T) {
 
 		err = await.WaitForCheInstallConditions(cheInstallation.Name, UntilHasCheStatusCondition(cheinstallation.InstallationSucceeded()))
 		require.NoError(t, err)
-		checkCheResources(t, f.Client.Client, cheOperatorNS, cheOg, cheSub, cheCluster)
+		cluster, err := await.GetCheCluster(cheCluster.Namespace, cheCluster.Name)
+		require.NoError(t, err)
```
Comment (MatousJobanek) [inline test/e2e/toolchain_test.go]: Why do you need to get the CheCluster resource - shouldn't that be loaded by the `CheClusterAssertion`?

File: test/e2e/toolchain_test.go [hunk 9]
```diff
@@ -54,29 +57,33 @@ func TestToolchain(t *testing.T) {
...
 
-	// 	err = await.WaitForCheInstallConditions(cheInstallation.Name, UntilHasCheStatusCondition(cheinstallation.SubscriptionCreated()))
-	// 	require.NoError(t, err)
-	// 	checkCheResources(t, f.Client.Client, cheOperatorNS, cheOg, cheSub)
-	// })
+		err = await.WaitForCheInstallConditions(cheInstallation.Name, UntilHasCheStatusCondition(cheinstallation.InstallationSucceeded()))
+		require.NoError(t, err)
+		cluster, err := await.GetCheCluster(cheCluster.Namespace, cheCluster.Name)
+		require.NoError(t, err)
```
Comment (MatousJobanek) [inline test/e2e/toolchain_test.go]: same here

File: test/e2e/toolchain_test.go [hunk 10]
```diff
@@ -125,7 +132,9 @@ func TestToolchain(t *testing.T) {
 
 		err = await.WaitForCheInstallConditions(cheInstallation.Name, UntilHasCheStatusCondition(cheinstallation.InstallationSucceeded()))
 		require.NoError(t, err)
-		checkCheResources(t, f.Client.Client, cheOperatorNS, cheOg, cheSub, cheCluster)
+		cluster, err = await.GetCheCluster(cheCluster.Namespace, cheCluster.Name)
+		require.NoError(t, err)
```
Comment (MatousJobanek) [inline test/e2e/toolchain_test.go]: and here

File: pkg/controller/cheinstallation/cheinstallation_controller.go (hunk 3 above)
Comment (MatousJobanek) [reply pkg/controller/cheinstallation/cheinstallation_controller.go]: ok, now I understand:

> Also, when the namespace is in 'Terminating' state (if
an admin deleted it manually), we delete the CheCluster
for the same reason.

but isn't that trying to fix someone's else error? I mean, from now on it shouldn't happen that the namespace is in `Terminating` state because we didn't do the proper cleanup.
In addition, if the operator is uninstalled and the `CheCluster` is not removed (because it contains finalizer) then calling another `delete` for that resource will not help anyhow. Or am I missing something?

File: make/test.mk [hunk 11]
```diff
@@ -153,8 +153,9 @@ endif
...
-	oc get subscription --output=name -n ${TOOLCHAIN_NS} |  grep "toolchain-operator" | xargs oc delete -n ${TOOLCHAIN_NS} 2>/dev/null || true
-	oc get subscription --output=name -n openshift-operators |  grep "openshift-pipelines-operator" | xargs oc delete -n openshift-operators 2>/dev/null || true
-	oc get csv --output=name -n openshift-operators |  grep "codeready-toolchain-operator" | xargs oc delete -n openshift-operators 2>/dev/null || true
+	-oc get catalogsource --output=name -n openshift-marketplace | grep "toolchain-operator" | xargs oc delete -n openshift-marketplace 2>/dev/null || true
+	-oc get subscription --output=name -n ${TOOLCHAIN_NS} | grep "toolchain-operator" | xargs oc delete -n ${TOOLCHAIN_NS} 2>/dev/null || true
+	-oc get subscription --output=name -n openshift-operators | grep "openshift-pipelines-operator" | xargs oc delete -n openshift-operators 2>/dev/null || true
+	-oc get csv --output=name -n openshift-operators | grep "codeready-toolchain-operator" | xargs oc delete -n openshift-operators 2>/dev/null || true
+	-oc get clusterrole --output=name  | grep "codeready-toolchain-operator" | xargs oc delete 2>/dev/null || true
```
[the remark this answers is not in the document]
Comment (xcoulon) [reply make/test.mk]: right, I should probably remove the `|| true` as well, since the `-` prefix is an idiom of Makefile.

File: pkg/controller/cheinstallation/cheinstallation_controller.go (hunk 2 above)
Comment (xcoulon) [reply pkg/controller/cheinstallation/cheinstallation_controller.go]: because when the request comes from the CheCluster resource which is watched, the `request.Namespace` argument to get the `CheInstallation` resource is incorrect

File: make/test.mk (hunk 11 above)
[the remark this answers is not in the document]
Comment (MatousJobanek) [reply make/test.mk]: well, I added the suffix `|| true` because the output looks better, so I would keep the suffix and remove the prefix.
Or - after all - you can keep them both... :-)

File: pkg/controller/cheinstallation/cheinstallation_controller.go [hunk 12]
```diff
@@ -109,14 +117,37 @@ func (r *ReconcileCheInstallation) Reconcile(request reconcile.Request) (reconci
...
 		}
 		return reconcile.Result{}, err
 	}
+	// ensure there's a finalizer, unless it's being deleted
+	if !util.IsBeingDeleted(cheInstallation) {
+		// Add the finalizer if it is not present
+		if err := r.addFinalizer(reqLogger, cheInstallation); err != nil {
+			return reconcile.Result{}, err
+		}
+	} else if util.IsBeingDeleted(cheInstallation) && util.HasFinalizer(cheInstallation, toolchainv1alpha1.FinalizerName) { // Che Installation is being deleted, but before that we should delete the Che Operator namespace explicitely
```
[the remark this answers is not in the document]
Comment (xcoulon) [reply pkg/controller/cheinstallation/cheinstallation_controller.go]: right, that should actually not be necessary

File: pkg/controller/cheinstallation/cheinstallation_controller.go (hunk 2 above)
Comment (MatousJobanek) [reply pkg/controller/cheinstallation/cheinstallation_controller.go]: gotcha - thanks for explanation :+1:

File: pkg/controller/cheinstallation/cheinstallation_controller.go (hunk 12 above)
[the remark this answers is not in the document]
Comment (xcoulon) [reply pkg/controller/cheinstallation/cheinstallation_controller.go]: changed in 8f6b363

File: make/test.mk (hunk 11 above)
[the remark this answers is not in the document]
Comment (xcoulon) [reply make/test.mk]: well, the `-` prefix is a feature of `Makefile` so I would rather keep it. But we can have both, indeed. So I'll just keep the change as-is for now, then ;)

[the remark this answers is not in the document]
Comment (MatousJobanek) [reply make/test.mk]: ok, agree - sorry for this unnecessary discussion :-D

File: pkg/controller/cheinstallation/cheinstallation_controller.go (hunk 3 above)
Comment (xcoulon) [reply pkg/controller/cheinstallation/cheinstallation_controller.go]: well, yes, it's aimed at making sure that the namespace is not stuck in `Terminating` phase if an admin deletes the namespace (which will be recreated by our operator afterwards). It does not cover all cases for sure, but at least this one will work.

File: pkg/controller/cheinstallation/cheinstallation_controller.go (hunk 4 above)
Comment (xcoulon) [reply pkg/controller/cheinstallation/cheinstallation_controller.go]: yes :)

Comment (xcoulon) [reply pkg/controller/cheinstallation/cheinstallation_controller.go]: removed in 13c4f03

File: make/test.mk (hunk 11 above)
[the remark this answers is not in the document]
Comment (xcoulon) [reply make/test.mk]: no worries ;)

File: pkg/controller/cheinstallation/cheinstallation_controller.go (hunk 2 above)
Comment (xcoulon) [reply pkg/controller/cheinstallation/cheinstallation_controller.go]: my pleasure ;)

File: pkg/controller/cheinstallation/cheinstallation_controller.go (hunk 1 above)
Comment (xcoulon) [reply pkg/controller/cheinstallation/cheinstallation_controller.go]: right, let me change that

File: pkg/controller/cheinstallation/cheinstallation_controller.go (hunk 3 above)
Comment (xcoulon) [reply pkg/controller/cheinstallation/cheinstallation_controller.go]: ok, I got your point, and you're right indeed! I reverted this change in f40f0f9

File: pkg/controller/cheinstallation/cheinstallation_controller.go (hunk 1 above)
Comment (xcoulon) [reply pkg/controller/cheinstallation/cheinstallation_controller.go]: done in f40f0f9

File: test/e2e/toolchain_test.go (hunk 8 above)
Comment (xcoulon) [reply test/e2e/toolchain_test.go]: at this stage of the e2e test, the initial `cheCluster` obj is not up-to-date so I need to explicitly get it again in order to compare its `status.CheURL` with what's in the `CheInstallation` resource.

File: test/e2e/toolchain_test.go (hunk 9 above)
Comment (xcoulon) [reply test/e2e/toolchain_test.go]: same answer ;)

File: test/e2e/toolchain_test.go (hunk 10 above)
Comment (xcoulon) [reply test/e2e/toolchain_test.go]: same answer as well ;)

File: pkg/controller/cheinstallation/cheinstallation_controller_test.go (hunk 6 above)
Comment (xcoulon) [reply pkg/controller/cheinstallation/cheinstallation_controller_test.go]: yeah, right. My intent was to make sure that the CheCluster resource was "marked for deletion" (assuming that in a real case, it would be processed by the Che operator afterwards). But we could also just check that it's deleted and not available in the client cache at the end of the test, indeed. 

And yes, I will also verify that the CheInstallation resource has no finalizer. Good point!

Comment (xcoulon) [reply pkg/controller/cheinstallation/cheinstallation_controller_test.go]: done in 6030df0

File: test/e2e/toolchain_test.go (hunk 8 above)
Comment (MatousJobanek) [reply test/e2e/toolchain_test.go]: Sure, but the [loadCheClusterAssertion()](https://github.com/codeready-toolchain/toolchain-operator/pull/62/files#diff-b11ae265569f1cc9752c7c327b74b47cR38) function should ensure that it loads the latest version of the resource. 
After all, you fixed this behavior a few days ago: https://github.com/codeready-toolchain/toolchain-operator/commit/221b8bef254238e8ca4d2bb154d20286aaa90208#diff-b11ae265569f1cc9752c7c327b74b47c

File: pkg/controller/cheinstallation/cheinstallation_controller_test.go (hunk 7 above)
Comment (xcoulon) [reply pkg/controller/cheinstallation/cheinstallation_controller_test.go]: good idea! done in 1d2da80

File: pkg/controller/cheinstallation/cheinstallation_controller.go (hunk 12 above)
[the remark this answers is not in the document]
Comment (MatousJobanek) [reply pkg/controller/cheinstallation/cheinstallation_controller.go]: could you please also fix the comment at the end of the line - as I suggested in my previous comment?

File: test/e2e/toolchain_test.go (hunk 8 above)
Comment (xcoulon) [reply test/e2e/toolchain_test.go]: good point. Let me try without this explicit loading.

Comment (xcoulon) [reply test/e2e/toolchain_test.go]: ok, so I removed this unnecessary loading in e9cfc39

File: test/e2e/toolchain_test.go (hunk 9 above)
Comment (xcoulon) [reply test/e2e/toolchain_test.go]: same here: reverted in cf0fe10

File: test/e2e/toolchain_test.go (hunk 10 above)
Comment (xcoulon) [reply test/e2e/toolchain_test.go]: same here: reverted in cf0fe10

File: pkg/controller/cheinstallation/cheinstallation_controller.go (hunk 12 above)
[the remark this answers is not in the document]
Comment (xcoulon) [reply pkg/controller/cheinstallation/cheinstallation_controller.go]: sure, sorry for missing that. Done in 8dc517a

File: pkg/controller/cheinstallation/cheinstallation_controller.go [hunk 13]
```diff
@@ -162,13 +210,12 @@ func (r *ReconcileCheInstallation) ensureCheNamespace(logger logr.Logger, cheIns
...
-			ns := v1.Namespace{}
+			ns := corev1.Namespace{}
 			if err := r.client.Get(context.TODO(), types.NamespacedName{Name: cheOpNamespace}, &ns); err != nil {
 				return false, err
 			}
-			if ns.Status.Phase != v1.NamespaceActive {
-				logger.Info("Namespace is not in active state", "namespace", ns.Name, "phase", ns.Status.Phase)
-				return true, nil // requeue until the namespace is active
+			if ns.Status.Phase != corev1.NamespaceActive {
+				return true, nil // requeue until namespace is active (again)
```
Comment (MatousJobanek) [inline pkg/controller/cheinstallation/cheinstallation_controller.go]: I would keep the info-log there - so we know what the controller "is waiting" for
```suggestion
				logger.Info("Namespace is not in active state", "namespace", ns.Name, "phase", ns.Status.Phase)
				return true, nil // requeue until the namespace is active
```

File: pkg/controller/cheinstallation/cheinstallation_controller_test.go [hunk 14]
```diff
@@ -475,6 +482,49 @@ func TestReconcile(t *testing.T) {
...
+			_, err = r.Reconcile(request)
+			// then
+			require.NoError(t, err)
+			// in that case, expect CheInstallation to have no finalizers
+			updatedCheInstallation := v1alpha1.CheInstallation{}
+			err = r.client.Get(context.TODO(), types.NamespacedName{
+				Name: cheInstallation.Name,
+			}, &updatedCheInstallation)
+			require.NoError(t, err)
+			assert.Empty(t, updatedCheInstallation.GetFinalizers())
```
Comment (MatousJobanek) [inline pkg/controller/cheinstallation/cheinstallation_controller_test.go]: just a detail - wouldn't it be easier to add a new method `HasNoFinalizer` to the `CheInstallationAssertion` type? and call the assertion here?

File: pkg/controller/cheinstallation/cheinstallation_controller_test.go (hunk 7 above)
Comment (MatousJobanek) [reply pkg/controller/cheinstallation/cheinstallation_controller_test.go]: I meant to add it to all the test-cases when `CheCluster` exists (in the same way as we always check that the `CheCluster` exists) - one can easily get lost as part of any refactoring effort. 

On the other hand, it's probably fine to have only one check there - so it's up to you ;-)

File: pkg/controller/cheinstallation/cheinstallation_controller.go (hunk 13 above)
Comment (xcoulon) [reply pkg/controller/cheinstallation/cheinstallation_controller.go]: ok, let me revert that

File: pkg/controller/cheinstallation/cheinstallation_controller_test.go (hunk 14 above)
Comment (xcoulon) [reply pkg/controller/cheinstallation/cheinstallation_controller_test.go]: good idea! Done f537f76

**Comment by xcoulon** [comment]
> Good job @xcoulon
> I have only two comments, but nothing significant

I added more assertions on the CheInstallation conditions and finalizers in cd859cf
